common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,249 @@
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.watchedFilePath);
+
+            isWatchRegistered = true;
+
+            while (!Thread.currentThread().isInterrupted()) {
+                try {
+                    watchKey = retrieveWatchKeyFrom(watchService);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    logException(e);
+                    // Allow thread to exit.
+                }
+
+                if (watchKey != null) {
+                    Optional<WatchEvent<?>> modificationDetectionEvent = watchKey.pollEvents()
+                            .stream()
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .findAny();
+
+                    if (modificationDetectionEvent.isPresent()) {
+                        log.info("Detected that the file [{}] has modified", this.watchedFilePath);
+                        callback.accept(modificationDetectionEvent.get());
+                    }
+
+                    boolean isKeyValid = watchKey.reset();
+                    log.debug("Done resetting watch key.");
+                    if (!isKeyValid) {
+                        log.info("No longer watching file [{}]", this.watchedFilePath);
+                        break;
+                    }
+                } else {
+                    log.debug("watchKey for file at path {} was null", this.watchedFilePath);
+                }
+                if (!loopContinuously) {
+                    break;
+                }
+            }
+        } catch (IOException e) {
+            logException(e);
+            throw new RuntimeException(e);
+        } finally {
+            if (watchKey != null) {
+                watchKey.cancel();
+            }
+            if (watchService != null) {
+                try {
+                    watchService.close();
+                } catch (IOException e) {
+                    log.warn("Error closing watch service", e);
+                }
+            }
+        }
+        log.info("Thread [{}], watching for modifications in file [{}] exiting,", getName(), this.watchedFilePath);
+    }
+
+    private WatchKey retrieveWatchKeyFrom(WatchService watchService) throws InterruptedException {
+        WatchKey result = watchService.take();
+        log.info("Retrieved and removed watch key for watching file at path: {}", this.watchedFilePath);
+
+        // Each file modification/create usually results in the WatcherService reporting the WatchEvent twice,
```

Comment:
Do we know what kind of modification it has been?  We only care about content modification, we don't care about time stamp updates, and if we can differentiate the updates, then that's a better approach than adding a sleep. In the case it needs to be a sleep call, it is not clear why it needs to be 200ms. We don't expect frequent changes, so maybe it can be longer, although it clearly affects creation time.

Reply:
The 200 ms is a rather arbitrary value. A few examples I have seen used 50 or 100 ms. I kept 200 ms. just to make it more likely that the events will get deduplicated. I think 200 ms. is OK, based on all the testing I have done so far on Kubernetes clusters with Cert Manager.

Reply:
You've only answered the question about the sleep time, but the first part of the comment is more important for me because if that's possible, then we can eliminate this sleep altogether.

Reply:
Sure. Each file modification/create usually results in the WatcherService reporting the WatchEvent twice, as the file is updated twice: once for the content and once for the file modification time.  These events occur in quick succession. We should be able to differentiate the updates AFAIK, but then we'd need to keep track of the events in a stateful manner, which would make the code more complex. In most cases, the current deduplication logic should have the desired consequences; In the rare case where the two events are tackled nonetheless, there is not much downside except for creating an extra SSL Context and assigning it to the shared variable. Therefore, I think the current approach is a reasonable choice. 

Hope that answers your question.

Reply:
> We should be able to differentiate the updates AFAIK

If you are able to differentiate the type of change, then it is as simple as discarding whatever is not a content modification. In your argument, you talk about (1) content change and (2) timestamp change. If we are able to differentiate between (1) and (2), then we can simply discard (2) and only process (1). Why can't we do it this way and eliminate the sleep? This is sleep is causing unnecessary confusion in the code.

Reply:
@fpj I had misunderstood your question earlier. I meant to say previously that the two events - one for the content change and one for the timestamp update - are two separate events. We only get a `WatchEvent` that has a method `context()` which gives us the "relative path between the directory registered with the watch service, and the entry that is created, deleted, or modified"  ([src](https://docs.oracle.com/javase/8/docs/api/java/nio/file/WatchEvent.html#context--)). So, we can't distinguish between the two events except by indirect interpretation based on the time that they have arrived (stateful processing). Hope that clarifies.

Reply:
At the end of the day, a hash of the content would suffice to distinguish a substantive update.  That's what the istio reference code that I shared does.

Reply:
I think what we have is sufficient for now. It works and does that without having to read the file based on JVM generated events alone. For cert manager distributed certificates we use the polling monitor instead of this class. I suggest we enhance this class later if needed.